### PR TITLE
HeanCMS: Add fetchStrategy

### PR DESCRIPTION
**Pull request by bapeey**
Checklist:

- [ ] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [x] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio

### CODE REVIEW

File: multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/heancms/HeanCms.kt [hunk 1]
```diff
@@ -230,21 +260,45 @@ abstract class HeanCms(
     }
 
     override fun mangaDetailsParse(response: Response): SManga {
-        val result = runCatching { response.parseAs<HeanCmsSeriesDto>() }
-        val seriesDetails = result.getOrNull()?.toSManga(apiUrl, coverPath, fetchAllTitles)
+        var mangaResponse = response
+        if (fetchAllTitlesStrategy == FetchStrategy.SEARCH_EACH) {
+            mangaResponse = getNewSlugResponse(response)
+        }
```
Comment (AwkwardPeak7) [inline multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/heancms/HeanCms.kt:266]:
```suggestion
        val mangaResponse = if (fetchAllTitlesStrategy == FetchStrategy.SEARCH_EACH) {
            getNewSlugResponse(response)
        else {
            response
        }
```

File: multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/heancms/HeanCms.kt [hunk 2]
```diff
@@ -230,21 +260,45 @@ abstract class HeanCms(
...
 
     override fun chapterListRequest(manga: SManga): Request = mangaDetailsRequest(manga)
 
     override fun chapterListParse(response: Response): List<SChapter> {
-        val result = response.parseAs<HeanCmsSeriesDto>()
-        val seriesSlug = response.request.url.pathSegments.last()
+        var mangaResponse = response
+        if (fetchAllTitlesStrategy == FetchStrategy.SEARCH_EACH) {
+            mangaResponse = getNewSlugResponse(response)
+        }
```
Comment (AwkwardPeak7) [inline multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/heancms/HeanCms.kt:291]:
```suggestion
        val mangaResponse = if (fetchAllTitlesStrategy == FetchStrategy.SEARCH_EACH) {
            getNewSlugResponse(response)
        else {
            response
        }
```

File: multisrc/overrides/heancms/yugenmangas/src/YugenMangas.kt [hunk 3]
```diff
@@ -45,26 +33,6 @@ class YugenMangas :
         timeZone = TimeZone.getTimeZone("UTC")
     }
 
-    private fun getFetchAllSeriesPref(): Boolean {
-        return preferences.getBoolean(FETCH_ALL_SERIES_PREF, FETCH_ALL_SERIES_DEFAULT_VALUE)
-    }
-
-    override fun setupPreferenceScreen(screen: PreferenceScreen) {
-        val fetchAllSeriesPreference = SwitchPreferenceCompat(screen.context).apply {
-            key = FETCH_ALL_SERIES_PREF
```
Comment (AwkwardPeak7) [inline multisrc/overrides/heancms/yugenmangas/src/YugenMangas.kt:54]: also remove these constants from companion object

File: multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/heancms/HeanCms.kt [hunk 4]
```diff
@@ -230,21 +260,47 @@ abstract class HeanCms(
...
+        } else {
+            response
+        }
+        val result = mangaResponse.parseAs<HeanCmsSeriesDto>()
+
+        if (fetchAllTitlesStrategy == FetchStrategy.SEARCH_EACH) {
+            val slugMap = getSlugMap().toMutableMap()
+            val permSlug = result.slug.substringAfterLast("/").replace(TIMESTAMP_REGEX, "")
+            slugMap[permSlug] = result.slug.substringAfterLast("/")
+            putSlugMap(slugMap)
```
Comment (alessandrojean) [inline multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/heancms/HeanCms.kt:300]:
```suggestion
            val permSlug = result.slug
                .substringAfterLast("/")
                .replace(TIMESTAMP_REGEX, "")
            preferences.slugMap = preferences.slugMap.toMutableMap()
                .also { it[permSlug] = result.slug.substringAfterLast("/") }
```

File: multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/heancms/HeanCms.kt [hunk 5]
```diff
@@ -230,30 +260,72 @@ abstract class HeanCms(
...
+            .find { it.replace(TIMESTAMP_REGEX, "") == originalSlug }
+            ?: throw Exception(intl.urlChangedError(name))
+
+        val apiHeaders = headersBuilder()
+            .add("Accept", ACCEPT_JSON)
+            .build()
+
+        val jsonString = client.newCall(GET("$apiUrl/series/$mangaSlug", apiHeaders)).execute().body.string()
+
+        val result = jsonString.parseAs<HeanCmsSeriesDto>()
```
Comment (alessandrojean) [inline multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/heancms/HeanCms.kt:317]: Can use the `Response.parseAs` extension function directly.

```suggestion
        val detailsRequest = GET("$apiUrl/series/$mangaSlug", apiHeaders)
        val result = client.newCall(detailsRequest).execute()
            .parseAs<HeanCmsSeriesDto>()
```

**Review by alessandrojean — APPROVED** [review]
Seems much better now 👍🏻, just this minor suggestion, but feel free to ignore it as well if you prefer.

**Pull request merged**